test/test_main.py

```diff
@@ -305,18 +305,23 @@ class TestCharmEvents(ops.CharmEvents):
 
         def cleanup():
             ops.CharmBase.on = ops.CharmEvents()  # type: ignore
-        self.addCleanup(cleanup)
+        request.addfinalizer(cleanup)
 
-        fake_script(typing.cast(unittest.TestCase, self), 'is-leader', 'echo true')
-        fake_script(typing.cast(unittest.TestCase, self), 'juju-log', 'exit 0')
+        fake_script.write('is-leader', 'echo true')
+        fake_script.write('juju-log', 'exit 0')
 
         # set to something other than None for tests that care
         self.stdout = None
         self.stderr = None
 
-    def _setup_charm_dir(self):
+    def _setup_charm_dir(self, request: pytest.FixtureRequest):
         self._tmpdir = Path(tempfile.mkdtemp(prefix='tmp-ops-test-')).resolve()
-        self.addCleanup(shutil.rmtree, str(self._tmpdir))
+
+        def cleanup():
+            shutil.rmtree(self._tmpdir) if self._tmpdir.exists() else None
```

Comment:
It's because there are a few test cases where they call the method in a for loop, for example, [here](https://github.com/canonical/operator/blob/main/test/test_main.py#L338).

The teardown will be triggered multiple times for the same test case, since there are multiple finalizers called, and the one called after the first time will error on this. I do not know why unittest `addCleanup` works, and I've also found that if I use `functools.partial` in finalizer, it works as well, but `shutil.rmtree` doesn't. I also tried `tmp_path` fixture, same story. That's how I got here 😆

Reply:
`shutil.rmtree(self._tmpdir, ignore_errors=True)` would be simpler, although that ignore a broader set of errors.

For the loop case, it seems like it might be better if the creation of the temp directory was separated out from populating it, so that the code created it once (and at that point added the cleanup) and then could populate it as many times as it likes (no extra cleanup required).

Reply:
For minimal change, I will use `shutil.rmtree(self._tmpdir, ignore_errors=True)` for now.